repo2docker/buildpacks/julia/julia_semver.py

```diff
@@ -0,0 +1,119 @@
+    return None
+
+def str_to_version(vstr):
+    return tuple([int(n) for n in vstr.split('.')])
+
+# --- Matcher interface -------------------------------------------
+
+class AbstractMatcher:
+    def match(self, v):
+        pass
+
+class SemverMatcher(AbstractMatcher):
+    """ Match a version tuple to a given constraint_str using the `semver` package. """
+    def __init__(self, constraint_str):
+        self.constraint_str = constraint_str
+    def match(self, v):
+        while len(v) < 3:
+            v = v+(0,)
+        v_str = '.'.join(map(str, v))
+        return semver.match(v_str, self.constraint_str)
+    def __repr__(self):
+        return self.constraint_str
+
+# --- Custom matcher for julia-specific SemVer handling: ---------
+
+from enum import Enum
+class Exclusivity(Enum):
+    EXCLUSIVE = 1
+    INCLUSIVE = 2
+
+class VersionRange(AbstractMatcher):
+    """ Match a version tuple between lower and upper bounds. """
+    def __init__(self, lower, upper, upper_exclusivity):
+        self.lower = lower
+        self.upper = upper
+        self.upper_exclusivity = upper_exclusivity
+    def match(self, v):
+        if self.upper_exclusivity == Exclusivity.EXCLUSIVE:
+            return self.lower <= v < self.upper
+        else:
+            return self.lower <= v <= self.upper
+    def __repr__(self):
+        return ("["+".".join(map(str, self.lower)) +"-"+ ".".join(map(str, self.upper)) +
+                    (")" if self.upper_exclusivity == Exclusivity.EXCLUSIVE else "]"))
+
+# Helpers
+def major(v): return v[0]
+def minor(v): return v[1] if len(v) >= 2 else 0
+def patch(v): return v[2] if len(v) >= 3 else 0
+
+# --- main constraint parser function ------------------------------------
+
+def create_semver_matcher(constraint_str):
+    """
+    Returns a derived-class instance of AbstractMatcher that matches version
+    tuples against the provided constraint_str.
+    """
+    constraint_str = constraint_str.strip()
+    first_digit = re.search(r"\d", constraint_str)
+    if not first_digit:
+        # Invalid version string (no numbers in it)
+        return ""
+    constraint = str_to_version(constraint_str[first_digit.start():])
+
+    comparison_symbol = constraint_str[0:first_digit.start()].strip()
+
+    # Default to "^" search if no matching mode specified (up to next major version)
+    if (first_digit.start() == 0) or (comparison_symbol == "^"):
+        if major(constraint) == 0:
+            # Also, julia treats pre-1.0 releases specially, as if the first
+            # non-zero number is actually a major number:
+            # https://docs.julialang.org/en/latest/stdlib/Pkg/#Caret-specifiers-1
+            # So we need to handle it separately by bumping the first non-zero number.
+            for i,n in enumerate(constraint):
+                if n != 0 or i == len(constraint)-1: # (using the last existing number handles situations like "^0.0" or "^0")
+                    upper = (*constraint[0:i], n+1)
```

Comment:
@betatim This syntax is apparently not supported on Python < 3.5... Any suggestion how we could write this instead? My Python is not exactly top notch :) And I assume we do need to support older Python versions, right?

Reply:
I think the python2 way to do this would be "tuple concatenation" via `+`?

I just tried it, and I _think_ this should work:
```python
upper = constraint[0:i] + (n+1,)
```

Reply:
(This is what i tried:
```python
>>> constraint = (0,6,1)
>>> i,n = (2,1)
>>> constraint[0:i] + (n+1,)
(0, 6, 2)
```
)

Reply:
Alright I tried that in the latest push, lets see whether it works!

Reply:
We go back to Python 3.4 in terms of supporting old versions.